### ISSUE TITLE
DegenerateBoxError on Training SSD 7 on Udacity Dataset

### ISSUE THREAD

**Issue by rodsnjr** [pierluigiferrari/ssd_keras#100]
Hello, I'm following the SSD 7 tutorial to your newly updated version ot the repository, and getting this error after the 1th epoch:

> ---------------------------------------------------------------------------
> DegenerateBoxError                        Traceback (most recent call last)
> <ipython-input-7-77c278ae603d> in <module>()
>       9                               validation_data=val_generator,
>      10                               verbose=0,
> ---> 11                               validation_steps=ceil(val_dataset_size/batch_size))
> 
> /home/rodsnjr/miniconda3/lib/python3.6/site-packages/keras/legacy/interfaces.py in wrapper(*args, **kwargs)
>      89                 warnings.warn('Update your `' + object_name +
>      90                               '` call to the Keras 2 API: ' + signature, stacklevel=2)
> ---> 91             return func(*args, **kwargs)
>      92         wrapper._original_function = func
>      93         return wrapper
> 
> /home/rodsnjr/miniconda3/lib/python3.6/site-packages/keras/engine/training.py in fit_generator(self, generator, steps_per_epoch, epochs, verbose, callbacks, validation_data, validation_steps, class_weight, max_queue_size, workers, use_multiprocessing, shuffle, initial_epoch)
>    2193                                 validation_generator,
>    2194                                 validation_steps,
> -> 2195                                 workers=0)
>    2196                         else:
>    2197                             # No need for try/except because
> 
> /home/rodsnjr/miniconda3/lib/python3.6/site-packages/keras/legacy/interfaces.py in wrapper(*args, **kwargs)
>      89                 warnings.warn('Update your `' + object_name +
>      90                               '` call to the Keras 2 API: ' + signature, stacklevel=2)
> ---> 91             return func(*args, **kwargs)
>      92         wrapper._original_function = func
>      93         return wrapper
> 
> /home/rodsnjr/miniconda3/lib/python3.6/site-packages/keras/engine/training.py in evaluate_generator(self, generator, steps, max_queue_size, workers, use_multiprocessing)
>    2308 
>    2309             while steps_done < steps:
> -> 2310                 generator_output = next(output_generator)
>    2311                 if not hasattr(generator_output, '__len__'):
>    2312                     raise ValueError('Output of generator should be a tuple '
> 
> /home/rodsnjr/miniconda3/lib/python3.6/site-packages/keras/utils/data_utils.py in get(self)
>     768             success, value = self.queue.get()
>     769             if not success:
> --> 770                 six.reraise(value.__class__, value, value.__traceback__)
> 
> /home/rodsnjr/miniconda3/lib/python3.6/site-packages/six.py in reraise(tp, value, tb)
>     691             if value.__traceback__ is not tb:
>     692                 raise value.with_traceback(tb)
> --> 693             raise value
>     694         finally:
>     695             value = None
> 
> /home/rodsnjr/miniconda3/lib/python3.6/site-packages/keras/utils/data_utils.py in _data_generator_task(self)
>     633                             # => Serialize calls to
>     634                             # infinite iterator/generator's next() function
> --> 635                             generator_output = next(self._generator)
>     636                             self.queue.put((True, generator_output))
>     637                         else:
> 
> /media/rodsnjr/Files/Git/ssd_keras/data_generator/object_detection_2d_data_generator.py in generate(self, batch_size, shuffle, transformations, label_encoder, returns, keep_images_without_gt, degenerate_box_handling)
>     790                     batch_y_encoded, batch_matched_anchors = label_encoder(batch_y, diagnostics=True)
>     791                 else:
> --> 792                     batch_y_encoded = label_encoder(batch_y, diagnostics=False)
>     793 
>     794             #########################################################################################
> 
> /media/rodsnjr/Files/Git/ssd_keras/ssd_encoder_decoder/ssd_input_encoder.py in __call__(self, ground_truth_labels, diagnostics)
>     306                 raise DegenerateBoxError("SSDInputEncoder detected degenerate ground truth bounding boxes for batch item {} with bounding boxes {}, ".format(i, labels) +
>     307                                          "i.e. bounding boxes where xmax <= xmin and/or ymax <= ymin. Degenerate ground truth " +
> --> 308                                          "bounding boxes will lead to NaN errors during the training.")
>     309 
>     310             # Maybe normalize the box coordinates.
> 
> DegenerateBoxError: SSDInputEncoder detected degenerate ground truth bounding boxes for batch item 0 with bounding boxes [[   1.    0.   13.  154.  163.]
>  [   1.   35.   51.  149.  161.]
>  [   1.   46.   98.  140.  172.]
>  [   1.   93.  106.  145.  162.]
>  [   1.  101.  133.  144.  165.]
>  [   1.  132.  156.  142.  161.]
>  [   1.  140.  180.  143.  174.]
>  [   1.  188.  211.  142.  162.]
>  [   1.  204.  253.  142.  184.]
>  [   1.  398.  479.  142.  193.]
>  [   1.  438.  478.  143.  154.]], i.e. bounding boxes where xmax <= xmin and/or ymax <= ymin. Degenerate ground truth bounding boxes will lead to NaN errors during the training.

The code is the same as your provided notebook, is this a bug, or the notebook needs to be updated ?

Thanks!

**Comment by pierluigiferrari** [comment]
What dataset are you training on? The Udacity dataset? It looks like the order of the box coordinates is `(xmin, xmax, ymin, ymax)`, but `SSDInputEncoder` requires its input to be in the format `(xmin, ymin, xmax, ymax)`. As the error says, `SSDInputEncoder` is being fed degenerate boxes. For example, in some of the boxes in that list above, element 1 (expected to be `xmin`) is larger than element 3 (expected to be `xmax`), that's why it's throwing the error.

**Comment by rodsnjr** [comment]
Hello.

> Did you make any changes to the SSD7 training notebook with regard to any coordinate format arguments in any function?

Nope, I was just testing the new commits, to understand how things are working now.

**Comment by pierluigiferrari** [comment]
That's not possible. I've just run the SSD7 training notebook without making any changes and I don't get any errors.

I've just checked the annotations CSV file, the box coordinates that are listed in the error message in your original post above are indeed in the incorrect format `(xmin, xmax, ymin, ymax)` when they should be in the format `(xmin, ymin, xmax, ymax)`. This incorrect coordinate order doesn't happen if you use the SSD7 training notebook as is, so you must have changed something somewhere.

Another indication that you changed something somewhere is that the XML error you posted in #101 likely stems from the same issue. As I wrote in that issue, I can't reproduce that other error either.

Once again, if you post a link to the full notebook that you're using I can take a look at what's going on. Otherwise I can only suggest to pull the latest master again, open the SSD7 training notebook, and change only the filepaths and try again.
